server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -2412,12 +2421,12 @@ public void onFailure(Exception e) {
                 public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
                     logger.trace("[{}] successfully set safe repository generation to [{}]", metadata.name(), newGen);
                     cacheRepositoryData(newRepositoryData, version);
-                    threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.supply(listener, () -> {
+                    snapshotDeleteTaskRunner.enqueueTask(ActionListener.wrap(releasable -> {
```

Comment:
nit/suggestion (sorry Github won't let use its suggestion feature here):

```diff
diff --git a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
index dfdc09b696f..af969087bd5 100644
--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
@@ -2421,7 +2421,7 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
                 public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
                     logger.trace("[{}] successfully set safe repository generation to [{}]", metadata.name(), newGen);
                     cacheRepositoryData(newRepositoryData, version);
-                    snapshotDeleteTaskRunner.enqueueTask(ActionListener.wrap(releasable -> {
+                    snapshotDeleteTaskRunner.enqueueTask(listener.map(releasable -> {
                         // Delete all now outdated index files up to 1000 blobs back from the new generation.
                         // If there are more than 1000 dangling index-N cleanup functionality on repo delete will take care of them.
                         // Deleting one older than the current expectedGen is done for BwC reasons as older versions used to keep
@@ -2436,8 +2436,8 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
                         } catch (IOException e) {
                             logger.warn(() -> "Failed to clean up old index blobs from before [" + newGen + "]", e);
                         }
-                        listener.onResponse(newRepositoryData);
-                    }, listener::onFailure));
+                        return newRepositoryData;
+                    }));
                 }
             });
         }, listener::onFailure);
```

NB slightly different semantics, this means we release `releasable` (potentially starting another delete task) before completing the grouped listener. I think that's ok tho.